### ISSUE TITLE
M2: reduce measured metadata and result-transfer overhead

### ISSUE THREAD

**Issue by hsliuustc0106** [ThinkFlowLab/vllm-rlt#5]
## Precision requirements — 2026-09-12

For new work, **BF16 is the primary GPU inference, kernel-validation, quality, and performance target**. Use BF16 weights, ordinary activations, and KV storage; specify accumulation independently and retain FP32 in numerically sensitive reductions/intermediates. Full-model FP32 is an optional diagnostic/reference configuration, not the default prerequisite. See the [precision policy](https://github.com/hsliuustc0106/vllm-lt/blob/c9e646ace2d8198a52e522ea4d71129f1b143c58/docs/precision-policy.md).

BF16 profiling, matched optimization A/B, and quality data collection may proceed alongside the original Q1 reference diagnosis after each run's functional, finiteness, and ownership prerequisites pass. Report observations, reference fidelity, and qualification separately. Preserve exact cache/data-movement/RNG invariants and metadata-only A/B behavior; differing arithmetic paths use independently justified, predeclared BF16 numerical and decision/quality criteria. Universal bitwise or top-1 equality is not a blanket mixed-precision requirement, and near-tie exemptions cannot be invented after a run.

For dtype choices, follow explicit original Ouro paper guidance first, then the pinned release configuration and inference code. The [source audit](https://github.com/hsliuustc0106/vllm-lt/blob/c9e646ace2d8198a52e522ea4d71129f1b143c58/docs/paper-notes.md#original-ouro-paper-and-precision-guidance) distinguishes published FP32 operations from backend/project choices; the paper has no per-kernel inference recipe.

This amendment supersedes conflicting dtype/eligibility requirements below for successor work. All dated contracts, raw results, original failures, and historical acceptance decisions remain unchanged. New criteria need a separately frozen contract; no benchmark or qualification is claimed by this edit.

Required follow-up: versioned BF16 M2 harness support and a new matched BF16 baseline/candidate comparison with identical accumulation settings, exact selected tensors/populated KV/tokens/exits/RNG progression, profiles, controls, and a finite run budget. Preserve the historical FP32 acceptance; this issue tracks the outstanding BF16 comparison as well.

- [ ] **AC-M2-BF16 — Representative precision evidence.** Execute the frozen BF16 comparison and report numerical/behavioral gates, both timing pairs, control/resource gates, and any unresolved independent-reference limitations separately. The FP32 20–21% observation does not satisfy this requirement.

Parent roadmap: #2. Milestone **M2 — reduce host overhead**.

Dependencies: M1 benchmark/profiler #3 and the selected dtype's numerical contract from [Q1 #4](https://github.com/hsliuustc0106/vllm-lt/issues/4). New work targets BF16 with explicit accumulation settings. The open historical Q1 diagnosis does not block matched BF16 optimization A/B; qualification is reported separately.

## Problem and decision gate

Source inspection identifies repeated work: `KVCacheManager.write()` and `attend()` rebuild row metadata and device tensors for every physical layer, while coda sampling reads back one scalar per request. Ouro has 24 physical layers per recurrent traversal. These are candidates, not measured bottlenecks or established speedups.

Start only after #3 provides attribution and nominates a relevant cost. If attention or another component dominates, defer this milestone and follow that evidence. Record the selected candidate, target workloads, primary metric, and allowed regressions before implementation experiments. This issue may produce several small PRs, but does not authorize implementing every candidate together.

## Goals, constraints, and non-goals

Reduce demonstrated host preparation or result-transfer overhead while preserving model arithmetic, gate probabilities, token/depth histories, KV isolation, and request RNG progression. Preserve synchronous scheduling, full-lifetime KV reservation, full-depth prompt prefill, and LAST-EXITED propagation. Use one Ouro model and the #3 prepared environment and harness.

Exclude attention-kernel tuning, CUDA graphs, device-side routing, preemption, cache compression, changed batch policies, and public API expansion. Do not combine a memory-layout or accumulation-policy change with host-overhead work. For the BF16 metadata comparison, use the same BF16 precision policy on both sides. Fewer metadata tensors or synchronization points alone are structural results; an optimization claim additionally needs measured inference benefit.

## Proposed interface: prepare metadata once per traversal

The first metadata PR should create an internal descriptor once for each `model.recurrent()` invocation and reuse it across physical layers. A prefill stage invokes recurrence at four depths and therefore prepares four depth-specific descriptors. It cannot reuse one unchanged depth table across all four loops.

Illustrative private boundary; names may follow repository conventions:

```python
@dataclass(frozen=True)
class PreparedKVBatch:
    rows: tuple[ValidatedRow, ...]  # host allocation/depth/position references
    position_ids: torch.Tensor     # [N], integer, runner device
    block_tables: torch.Tensor     # [N, table_width], int32
    context_lengths: torch.Tensor  # [N], int32, position + 1
    write_blocks: torch.Tensor     # [N], integer
    write_offsets: torch.Tensor    # [N], integer

def prepare_batch(request_ids, depths, positions) -> PreparedKVBatch: ...
def _write_prepared(layer, batch, k, v) -> None: ...
def _attend_prepared(layer, batch, q) -> torch.Tensor: ...
```

The KV manager owns address resolution. The runner/model passes the resulting descriptor through the internal layer calls; model code must not inspect free lists or allocate pages. Reuse `position_ids` for RoPE instead of creating another equivalent tensor. Keep existing public write/attention validation behavior through thin adapters where existing callers require it; avoid maintaining two independent address-resolution algorithms.

`prepare_batch()` validates equal row counts, allocations, depths, positions, duplicate write addresses, device metadata shapes, and table width. It produces layer-independent addresses. `_write_prepared()` still checks layer and K/V tensor compatibility; `_attend_prepared()` still checks query compatibility and that the specific layer's required prefix is initialized. Written-position tracking remains per layer and advances only after that layer's write is issued. Do not remove initialization checks because metadata was prepared earlier.

Descriptors live for one synchronous traversal. Do not add a cross-request or cross-step descriptor cache, cache invalidation framework, padding contract, or pinned-memory pool in this PR. Allocation ownership must remain valid until its final consuming operation completes. Reject use after release/reallocation if a descriptor can escape the internal call boundary; otherwise make that lifetime explicit and prevent escape. Pending positions and partial pages must retain existing semantics.

This extraction preserves current tensor arithmetic and attention dispatch. It moves shared validation/address construction out of the layer loop; it does not move layer-specific validity transitions out of their execution order.

## Separate candidate: batch token-result transfer

Only select this after profiling shows per-request coda readbacks matter. Keep logits computation and per-request sampling unchanged, returning device token scalars from the private sampling helper:

```python
def _sample_device(logits: torch.Tensor, request: Request) -> torch.Tensor: ...
# execute(CODA): sample in the current request order, stack IDs, then one
# CPU transfer and conversion to the existing list[int] result contract.
```

Each stochastic request keeps its own generator, seed, and sequence of draws. Preserve temperature/top-k/top-p behavior and tie semantics. Do not replace per-request random streams with a batch generator or vectorized sampler without a separately reviewed equivalence contract. Validate mixed greedy/stochastic batches even though #3's timing fixtures are greedy.

Avoiding unused gate readback for wholly fixed-depth batches is a third, separately justified candidate. It requires an explicit runner/update result contract and instrumentation accounting. Adaptive gates before minimum exit depth still contribute to cumulative probability. #3 replay deliberately retains real gate work; keep that benchmark policy explicit rather than silently turning replay into a cheaper execution path. Gate computation or transfer removal is not required to complete the metadata PR.

## Validation and measurement contract

First run the CUDA-free CPU tests relevant to changed ownership and interfaces. Use the independent dense and serial comparisons retained by Q1; existing mixed-depth/cache tests remain mandatory. Metadata cases include fragmented allocations, adjacent tokens, partial blocks, multiple prefill chunks, heterogeneous depths/positions, cancellation, reuse, and invalid inputs. Sampling cases compare output IDs and subsequent generator progression under changed batching and completion order. Structural checks should verify shared preparation, not freeze incidental helper-call arrangements.

For device validation and A/B measurement, freeze source revisions/modification hashes and retain #3's workload identities, exact device IDs, model, dtype, arithmetic flags, residency, cache state, and timing boundaries. Predeclare the isolated change, target metric, maximum regressions, and stop conditions. Begin with a CPU probe and one excluded feasibility comparison; default to two measured runs per configuration after preparation/warmup, with any different budget declared in advance. Reserve all GPU work through the verified scheduler and clean only task-owned resources.

Report metadata CPU attribution, relevant allocations/transfers, and end-to-end latency/throughput separately. Preserve raw observations and ranges. Do not select a winner when the declared run budget leaves the difference within observed variation. A negligible original cost, unsupported attribution, correctness failure, exceeded regression budget, or absent representative benefit stops promotion of that candidate. Record a supported deferral rather than expanding scope automatically.

## Acceptance criteria / 验收标准

Performance targets are filled from the M1 baseline before candidate measurements: `P > 0` is required improvement, control regression limits are numeric and nonnegative, and resource/setup limits include units. Throughput improvement is `100 * (candidate / baseline - 1)`; latency improvement is `100 * (1 - candidate / baseline)`. Apply the selected definition to each matched pair and the corresponding control observations. A TBD target or a target chosen after seeing candidate results cannot pass acceptance.

Every checked AC must link its test log or immutable result artifact and implementation SHA. Record the selected candidate before implementation measurements. An unselected candidate is explicitly **out of scope**, not a passed test. AC-M2-01 through AC-M2-05 qualify correctness and structure; they do not demonstrate an optimization. Completion requires all applicable criteria, including AC-M2-06 through AC-M2-07.

- [ ] **AC-M2-01 — Evidence and frozen decision.** Link #3 attribution showing the selected metadata or transfer cost. Before candidate measurements, commit a contract specifying candidate/base SHAs and modification hashes, dtype/numerical contract, target workloads, primary metric, numeric improvement target `P%`, per-control regression limits, memory/setup budgets, comparison order, and variation rule. **Pass:** every field has a concrete value; artifact timestamps precede candidate measurements; controls differ only by the declared change and recorded necessary consequences. Pending BF16 qualification cannot be treated as passed FP32/BF16 equivalence.

- [ ] **AC-M2-02 — Metadata reuse and layer validity, if selected.** A bounded trace demonstrates one shared metadata preparation per recurrent traversal, including separate preparations for each prefill depth, with reuse across all 24 layers. CPU tests cover fragmented pages, partial blocks, heterogeneous depths/positions, chunk boundaries, duplicate addresses, missing history in an individual layer, and cancellation followed by page reuse. Compare complete populated KV prefixes against the retained reference. **Pass:** all comparisons and invalid-input checks pass; missing history still fails before attention; no stale descriptor changes recycled pages; structural evidence shows layer-independent construction without removing layer-specific validation.

- [ ] **AC-M2-03 — Sampling and transfer contract, if selected.** Compare baseline/candidate tokens, request generator states, and a subsequent draw for fixed seeds across mixed greedy/stochastic requests, differing completion order, and serial/refill/no-refill execution. A profiler capture counts the coda result transfers. **Pass:** exact token/draw-state agreement, one batch result transfer replacing per-request scalar transfers, and unchanged temperature/top-k/top-p/tie semantics. For any separately selected fixed-depth gate change, tests preserve adaptive cumulative hazards before minimum depth and #3's declared replay compute/readback work.

- [ ] **AC-M2-04 — Numerical and lifecycle preservation.** Run relevant CPU engine/cache tests and reserved per-loop hidden/gate/logit comparisons under the frozen dtype contract, plus complete generated token/depth histories. Include adjacent shallow/deeper exits, EOS, abort, execution failure and block reuse. **Pass:** every numerical comparison satisfies its predeclared tolerance; discrete histories agree exactly on the validation fixtures; all task requests and reserved request pages reach zero after safe cleanup. No failed dtype/shape case is dropped from reported coverage.

- [ ] **AC-M2-05 — Reviewable implementation.** Link the focused diff, passing required lint/format/tests, and compatibility evidence. **Pass:** public generation/output contracts and synchronous scheduling remain intact; no undeclared kernel, dtype, allocator or batch-policy change enters the comparison. Any missing required check leaves this criterion unchecked.

- [ ] **AC-M2-06 — Complete matched performance evidence.** After the CPU probe and excluded feasibility/warmup, run the frozen A–B/B–A comparison on identical exact devices using #3 timing boundaries. **Pass:** both pairs meet `P%` on every declared primary target; every required control observation respects its regression limit; memory/setup limits hold. Missing controls, a failing pair or averaged-away regression fails acceptance. If improvement is within observed variation under the frozen rule, mark it **inconclusive** at the run limit; do not add trials silently.

- [ ] **AC-M2-07 — Auditable promotion.** Publish raw runs, source/control hashes, both values/ranges, profiler attribution, memory/setup results, cleanup evidence and criterion-by-criterion outcomes. **Pass:** the evidence supports the implemented subset's benefit and all applicable gates are met. A deferral/no-go is a disposition, not accepted implementation: keep the issue open or explicitly close it as **not planned**, retaining unmet gates visibly unchecked.

Source snapshot: [KV metadata and allocation](https://github.com/hsliuustc0106/vllm-lt/blob/fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7/vllm_lt/core/kv_cache_manager.py), [runner and sampling](https://github.com/hsliuustc0106/vllm-lt/blob/fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7/vllm_lt/worker/model_runner.py).
